pkg/model/registry.go

```diff
@@ -293,48 +335,99 @@ func (r *Registry) isEqual(a, b v1.ObjectMeta) bool {
 
 // deleteEntity
 func (r *Registry) deleteEntity(entityType EntityType, meta v1.ObjectMeta) bool {
-	if r.Len() == 0 {
-		return false
+	// Try to minimize coarse grained locking at the registry level. Immediately getOrCreate for the entity type
+	// and then begin operating on that (it uses a finer grained lock).
+	setForType := r.getOrCreateObjSet(entityType)
+
+	// Having acquired the type-specific ObjectMeta set, return the result of removal success.
+	return setForType.remove(&meta)
+}
+
+// getOrCreateObjSet resolves the singleton objectMetaSet for this registry, of the given entityType.
+// This uses a coarse grained lock on the entire registry.
+func (r *Registry) getOrCreateObjSet(entityType EntityType) *objectMetaSet {
```

Comment:
We have an internal agreement that functions with behavior like 'get or create if not exists and then get' are called either `Ensure()` or starting with `Ensure`, like `Ensure<Something>()`
Ex.:
```
func (chi *ClickHouseInstallation) EnsureStatus() *ChiStatus {
```
or
```
func (o *StatefulSetPollOptions) Ensure() *StatefulSetPollOptions {
```
Most likely this is the case

Reply:
(Addressed)